### ISSUE TITLE
micro_speech example breaks when feeding it trained model on Arduino

### ISSUE THREAD

**Issue by leocorne** [tensorflow/tensorflow#39938]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- TensorFlow version (use command below):
1.15
- Microcontroller used:
Arduino Nano 33 BLE Sense (Arduino 1.8.12 IDE)

**Describe the current behavior**
micro_speech example works on pretrained model, but whenever I try to train a model and run it on the device it breaks (even with "yes" and "no" words). The serial port receives the following message:

Feature generation failed
Requested feature_data_ size 536907080 doesn't match 1960.

When running test_micro_speech_test with the same model it also crashes the program (it never stops running).

The model trains fine and achieves a good accuracy on colab and it doesn't give any errors when converting either.

**Describe the expected behavior**

It should run inference just like the pretrained model.

**Standalone code to reproduce the issue**
Trained with the following notebook (didn't alter anything):
https://colab.research.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/examples/micro_speech/train/train_micro_speech_model.ipynb

Using the Arduino TFLite library version 1.15

**Comment by leocorne** [comment]
It seems the newest version of the micro_speech training script creates a model that's incompatible with the TF Lite library downloadable in the Arduino IDE. Issue was fixed by building a .zip library with the latest version of TF Lite instead.

**Comment by jaredpb** [comment]
Hi do you mind telling me how you built the .zip library? I am a little new to Arduino and cloning the tensorflow source code into a zip file is not working for me.

**Comment by leocorne** [comment]
Hi. You have to build it with the following script (inside the TF GitHub
repo):

tensorflow/lite/micro/tools/ci_build/test_arduino.sh


It takes a few minutes but it will create the zip file at:

tensorflow/lite/micro/tools/make/gen/arduino_x86_64/prj/micro_speech/tensorflow_lite.zip

Hope that helps!

**Comment by jaredpb** [comment]
What's the simplest way to run shell commands on windows? I tried running it from powershell using sh test_arduino.sh but get a make error. I also tried to run the script on WSL but get a set illegal option error.

**Comment by leocorne** [comment]
I just use WSL. The Linux line endings and the windows ones are different
so it's causing you problems. This is the fix I used.

https://askubuntu.com/questions/966488/how-do-i-fix-r-command-not-found-errors-running-bash-scripts-in-wsl


Just apply dos2unix to the entire micro/tools folder and it will change the
line endings.

**Comment by jun66534** [comment]
@leocorne Hello. I ran test_arduino.sh and got tensorflow_lite.zip.
In example compling is success. but uploading failed.
What progammer did you use in Arduino IDE?

**Comment by jaredpb** [comment]
Hi leocorne, it looks like I still have one error getting in the way. When I am generating my zip I get a "No rule to make target 'tensorflow/lite/micro.../fixedpoint.h' needed by generate_hello_world_test_make_project'. Stop. " Any advice on what could be causing this?

**Comment by jun66534** [comment]
@leocorne  Hi, I upload the example in zip library. But it does't work.
I only delete two sparkfun_edge_* files. And I didn't train. Only  used example

**Comment by victorromeo** [comment]
I'm also still getting this error, using the latest Arduino Web Editor, but with my own model.  The model was created yesterday using TensorFlow 2.2.0 and the TFLite conversion done online using Google Colab.

I first received this error after using 'micro_speech' from the latest TensorFlow Arduino Library Zip (2.1.0 alpha), via a new clone of Tensorflow.

I appreciate that Arduino Web Editor is using TensorFlow Lite 1.15.0, however as I'm also getting this error using the latest source from the TensorFlow master branch, I don't think its related to the compile toolchain.

I performed the following steps on Arduino Web Editor, using my model:

- The model variable names were modified to use the original 'g_model' and 'g_model_len' variable names.
- Uploaded my model file, and accompanying .h file.
- The speech command word list was increased to include the new trained words and the CommandCount increased from 4 to 13
- micro_speech.ino file was modified to use the new model, replacing the old model.
![Screenshot from 2020-06-19 10-47-40](https://user-images.githubusercontent.com/3838993/85085508-6cf3ce80-b21b-11ea-96cf-77eb611c846e.png)

**Comment by victorromeo** [comment]
```
Feature generation failed
Requested feature_data_ size 536907080 doesn't match 1960
```

I have determined that the error message directly above is the typical behaviour exhibited by the loop() method, when the setup() method has failed to correctly run to completion.  In my case, as the pre-allocation of memory is insufficient in size for the Tensor Arena, or the model input data type is not correct.  There may other reasons also why this has failed.  As the serial monitor may not be connected during the setup() method being evaluated, the associated error messages may not be observed by the developer.

I found these faults in my application by converting the Arduino Web Editor sketch to a Platform IO project, then debugging on the device using a Segger JLink.  I am unaware of any other way to catch these errors at this time.

**Comment by ErickGQ1423** [comment]
Hello, 

I have the same problem and I can't solve it, could any one please explain me how you do it?
Feature generation failed
Requested feature_data_ size 536907080 doesn't match 1960.

**Comment by victorromeo** [comment]
@ErickGQ1423 I think you need to increase the size of the tensor arena.

**Comment by ErickGQ1423** [comment]
Thank you @victorromeo, did you solve the problem with this?

**Comment by victorromeo** [comment]
Yes. Recently, TFLite Arena Sizes have received additional support, by reporting the size of the actual tensor arena used.
```c++
// Guess a larger value of tensor arena 'arena_size' than is actually required.
...
// Build an interpreter to run the model with.
 tflite::MicroInterpreter interpreter(model, resolver, tensor_arena, arena_size, error_reporter);
...
 // This obtains the number of bytes that are actually in use
    return interpreter.arena_used_bytes();
```

Then change your code, to use the reported value + 100, to allow for future changes to the TFLite Framework implementation.

**Comment by ErickGQ1423** [comment]
Hello, @victorromeo could you please help me again?
I have been working for a week with the same problem, but I still have the same. I thing that I am probably not using the correct library in Arduino or probably another code on Colab, I have been following step by step that is in the book TinyML and it looks really easy, but I do not know why my code is not running

I hope you can help me.

**Comment by stanleyoz** [comment]
Why is this thread closed when the there is no resolution?

**Comment by seung-gu** [comment]
This thread shouldn't be closed, nobody brings any solution yet

**Comment by avisinghal6** [comment]
There is no resolution for this issue yet, it shouldnt be closed.

**Comment by avisinghal6** [comment]
I was able to solve the issue using the resolution at : https://github.com/tensorflow/tensorflow/issues/44868#issuecomment-845251763 

I cloned the latest version of tflite-mirco-examples, added it to the libraries in arduino IDE. Then after adding the lines as per the above resolution, I was able to run a new model on arduino. Also if we train the model using tensorflow v1.15, then  also I was able to run the model on arduino (i did not have to add the additional lines in this case.)
